gbpservice/nfp/modules/configurator.py

```diff
@@ -0,0 +1,458 @@
+
+from gbpservice.nfp.configurator.lib import constants
+from gbpservice.nfp.configurator.lib import demuxer
+from oslo_log import log
+from gbpservice.nfp.core import rpc
+from gbpservice.nfp.core import fifo as queue
+from gbpservice.nfp.configurator.lib import utils
+
+AGENTS_PKG = 'gbpservice.nfp.configurator.agents'
+CONFIGURATOR_RPC_TOPIC = 'configurator'
+LOG = log.getLogger(__name__)
+
+"""Implements procedure calls invoked by an REST server.
+
+Implements following RPC methods.
+  - create_network_device_config
+  - delete_network_device_config
+  - update_network_device_config
+  - create_network_service_config
+  - delete_network_service_config
+  - update_network_service_config
+  - get_notifications
+Also implements local methods for supporting RPC methods
+
+"""
+
+
+class ConfiguratorRpcManager(object):
+
+    def __init__(self, sc, cm, conf, demuxer):
+        self.sc = sc
+        self.cm = cm
+        self.conf = conf
+        self.demuxer = demuxer
+
+    def _get_service_agent_instance(self, service_type):
+        """Provides service agent instance based on service type.
+
+        :param service_type: firewall/vpn/loadbalancer/generic_config
+
+        Returns: Instance of service agent for a given service type
+
+        """
+
+        return self.cm.service_agent_instances[service_type]
+
+    def _send_request(self, operation, request_data):
+        """Maps and invokes an RPC call to a service agent method.
+
+        Takes help of de-multiplexer to get service type and corresponding
+        data and invokes the method of service agent. Service agent instance
+        is identified based on the service type passed in the request data
+
+        :param operation: Operation type - create/delete/update
+        :param request_data: RPC data
+
+        Returns: None
+
+        """
+
+        # Retrieves service type from RPC data
+        service_type = self.demuxer.get_service_type(request_data)
+        if (constants.invalid_service_type == service_type):
+            msg = ("Configurator received invalid service type %s." %
+                   service_type)
+            raise Exception(msg)
+
+        # Retrieves service agent information from RPC data
+        # Format of sa_info_list:
+        # [{'method': <m1>, 'kwargs': <rpc_data1>}, {}, ... ]
+        sa_info_list = self.demuxer.get_service_agent_info(
+                                                    operation,
+                                                    service_type,
+                                                    request_data)
+        if not sa_info_list:
+            msg = ("Configurator received invalid data format for service"
+                   " type %s. Data format: %r" % (service_type, request_data))
+            raise Exception(msg)
+
+        # Retrieves service agent instance using service type
+        sa_obj = self._get_service_agent_instance(service_type)
```

Comment:
- sa_instance instead of sa_obj
- Using service_agent in some cases and sa in some cases. Be consistent as much as poss